File: src/main/java/org/noxet/noxetserver/playerstate/properties/PSPHealth.java
package org.noxet.noxetserver.playerstate.properties;

import org.bukkit.configuration.ConfigurationSection;
import org.bukkit.entity.Player;
import org.bukkit.scheduler.BukkitRunnable;
import org.noxet.noxetserver.NoxetServer;
import org.noxet.noxetserver.playerstate.PlayerStateProperty;

public class PSPHealth implements PlayerStateProperty<Double> {
    @Override
    public String getConfigName() {
        return "health";
    }

    @Override
    public Double getDefaultSerializedProperty() {
        return 20D;
    }

    @Override
    public Double getSerializedPropertyFromPlayer(Player player) {
        return player.getHealth();
    }

    @Override
    public void restoreProperty(Player player, Double health) {
        new BukkitRunnable() {
            @Override
            public void run() {
                player.setHealth(health); // Restore health property on next tick, to prevent stuck outside of realm bug (let teleport happen first).
            }
        }.runTaskLater(NoxetServer.getPlugin(), 1);
    }

    @Override
    public Double getValueFromConfig(ConfigurationSection config) {
        return config.getDouble(getConfigName());
    }
}

File: src/main/java/org/noxet/noxetserver/playerstate/properties/PSPHealth.scala
package org.noxet.noxetserver.playerstate.properties

import org.bukkit.configuration.ConfigurationSection
import org.bukkit.entity.Player
import org.bukkit.scheduler.BukkitRunnable
import org.noxet.noxetserver.NoxetServer
import org.noxet.noxetserver.playerstate.PlayerStateProperty

object PSPHealth extends PlayerStateProperty[Double]:
  override def getConfigName: String = "health"

  override def getDefaultSerializedProperty: Double = 20

  override def getSerializedPropertyFromPlayer(player: Player): Double =
    player.getHealth

  override def restoreProperty(player: Player, health: Double): Unit =
    // Restore health property on next tick,
    // to prevent stuck outside of realm bug (let teleport happen first).
    QuickRunnable(player.setHealth(health))
      .runTaskLater(NoxetServer.getPlugin, 1)

  override def getValueFromConfig(config: ConfigurationSection): Double =
    config.getDouble(getConfigName)